calico/felix/test/test_fetcd_api.py

```diff
@@ -0,0 +1,177 @@
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(1)
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+        self.etcd_api.write_to_etcd.assert_has_calls([status_call,
+                                                      uptime_call])
+
+    def test_update_felix_status_continuous(self):
+        """
+        Test felix status is being continuously updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=5,
+                          REPORTING_TTL_SECS=17)
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(5)
+
+        last_status_call = None
+        for update in range(5):
+            gevent.sleep(5)
+            # Check that both uptime and status were updated
+            self.etcd_api.write_to_etcd.assert_has_calls([status_call, uptime_call])
+            # Update last_status_call and reset write_to_etcd calls
+            last_status_call = self.etcd_api.write_to_etcd.mock_calls[0]
+            self.etcd_api.write_to_etcd.reset_mock()
+
+    @patch('calico.felix.fetcd._reconnect')
+    def test_update_felix_status_reconnects_on_etcd_exception(self, _reconnect):
+        """
+        Test felix status handles exceptions
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=4,
+                          REPORTING_TTL_SECS=12)
+        self.etcd_api.write_to_etcd = Mock(side_effect=EtcdException)
+
+        gevent.sleep(1)
+
+        self.assertTrue(_reconnect.called)
+
+
+class TestIfStatus(object):
+    """
+    Used to check whether status has JSON format and contains expected
+    attributes, whis are:
+    - timestamp in ISO 8601 Zulu format
+    """
+    def __eq__(self, other):
+        import re
+        timestamp_regex = re.compile('.*"status_time": "\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z?".*')
+        has_timestamp =  timestamp_regex.match(str(other)) is not None
+        return is_json(other) and has_timestamp
+    def __repr__(self):
+        return '%s()' % self.__class__.__name__
+
+class TestIfUptime(object):
+    """
+    Used to check whether uptime has correct format (i.e. whether it is
+    non-negative integer)
+    """
+    def __eq__(self, other):
+        is_int = type(other) == int
+        is_non_negative = other >= 0
+        return is_int and is_non_negative
+    def __repr__(self):
+        return '%s()' % self.__class__.__name__
+
+def is_json(object):
+  try:
+    json.loads(object)
+  except ValueError, e:
+    return False
+  return True

```

Comment:
A few nitpicks:

* This function should have a docstring
* Indentation should be four spaces, not 2
* There's no need to capture the exception as `e` in `except ValueError, e:`; you can just do `except ValueError`.

  (And the preferred form for this construction is `except ValueError as e`. That form works with Python 2.6 to 3.5, but the comma syntax has been dropped in Python 3.)